lib/rspec/support/object_inspector.rb

```diff
@@ -3,23 +3,51 @@ module Support
     # Provide additional output details beyond what `inspect` provides when
     # printing Time, DateTime, or BigDecimal
     module ObjectInspector
+      DEFAULT_OPTIONS = {
+        :nested => false,
+        :joiner => ',',
+        :braces => true,
+        :trailing_comma => false
+      }
       # @api private
-      def self.inspect(object)
-        if Time === object
+      # rubocop:disable CyclomaticComplexity
+      # rubocop:disable MethodLength
+      def self.inspect(object, options=DEFAULT_OPTIONS)
+        case object
+        when Time
           format_time(object)
-        elsif defined?(DateTime) && DateTime === object
-          format_date_time(object)
-        elsif defined?(BigDecimal) && BigDecimal === object
-          "#{object.to_s 'F'} (#{object.inspect})"
-        elsif RSpec::Support.is_a_matcher?(object) && object.respond_to?(:description)
-          object.description
+        when Hash
+          format_hash(object, options)
```

Comment:
> Anyway, in order to keep the differ working, either this functionality has to be recreated within the ObjectInspector, or the differ class gets a bit of a rework to work without expecting the string passed in to have already been formatted by pp.

I think there's a 3rd option you haven't mentioned: `ObjectInspector` can replace items with their inspected equivalents and then delegate to `pp` to format the entire array rather than re-implementing that.  Then we don't have to concern our selves with the pretty print array formatting (pp takes care of that for us!) and stick to replacing objects with their better-output equivalents.  That should be pretty doable, right?

Anyhow, I'm realizing that our needs for diffing are significantly different from our needs for general object inspecting.  I really think we should make separate objects for these two cases.  Then one can focus on formatting of primitives and data structures for use in error messages and one can focus on formatting for diffing (which could subclass or delegate to the other inspector).  Among other things, I think that would allow us to remove the plethora of formatting options the `ObjectInspector` now takes.  Those options worry me for a few reasons:

* The point of the `ObjectInspector` is to take care of the formatting details so the caller doesn't have to, but by making options be part of the API we're moving some of the responsibility back to the caller.
* The options seem to be focused on the hash case, but the caller could be passing some other kind of object, in which case the options seem extraneous and weird.
* As I understand it, the options are there to support the differ vs non-differ case...but the options support _many_ combinations of values.  And yet, I don't think we want to support many combinations of values -- there's just two valid combinations we want to support, right?  Reifying those two sets of options as their own classes that implement their own logic seems like a step in the right direction.

Anyhow, if we do split the differ inspector from the general inspector I think it would make even more sense to handle arrays in the general inspector as part of this PR.  (The differ inspector could perhaps be pushed off to another PR if that helps limit scope here but it need not be).

As far as array/hash formatting for the general/non-differ inspector goes, you may want to follow the pattern we used for `RSpec::Expectations::Composable#surface_descriptions_in`:

* That method simply walks the data structure and replaces the items in it as appropriate with `DescribableItem` objects. ([source](https://github.com/rspec/rspec-expectations/blob/v3.2.1/lib/rspec/matchers/composable.rb#L99-L115)).
* The describable items then implement `inspect` and `pretty_print` so that if the array or hash is inspected or pretty printed the items inside use the desired description strings. ([source](https://github.com/rspec/rspec-expectations/blob/v3.2.1/lib/rspec/matchers/composable.rb#L172-L180))

I see that as being analogous for what we want for the non-diff case: we want to replace how time objects are inspected when they are nested in arrays/hashes, but not worry about the formatting of the array or hash itself.  It's nice to let ruby take care of that simply by using `Array#inspect` or `Hash#inspect` with the items in the data structures suitably replaced.

Sorry if this takes the PR in a different direction then what you had in mind, but I do think it will simplify and improve things.

Thoughts?